packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js

```diff
@@ -13,101 +13,522 @@
+  `);
+  await knex.raw(`
+    UPDATE pest_control_task
+    SET (weight, weight_unit) = (product_quantity, product_quantity_unit)
+    WHERE product_quantity IS NOT NULL
+    AND product_quantity_unit IN ${rawWeightUnits}
+  `);
+
+  // Product not required for pest control task
+  await knex.schema.alterTable('pest_control_task', (table) => {
+    table.check(
+      '(weight IS NULL AND volume IS NOT NULL) OR (volume IS NULL AND weight IS NOT NULL) OR (weight IS NULL AND volume IS NULL)',
+      [],
+      'volume_or_weight_check',
+    );
+    table.check(
+      '(weight IS NULL AND weight_unit IS NULL) OR (weight IS NOT NULL AND weight_unit IS NOT NULL)',
+      [],
+      'weight_unit_check',
+    );
+    table.check(
+      '(volume IS NULL AND volume_unit IS NULL) OR (volume IS NOT NULL AND volume_unit IS NOT NULL)',
+      [],
+      'volume_unit_check',
+    );
   });
 
-  await knex.raw(`
-      UPDATE soil_amendment_task
-      SET temp_purpose = purpose[1]
-    `);
+  // Alter cleaning task for volume and weight
+  await knex.schema.alterTable('cleaning_task', (table) => {
+    table.dropColumn('product_quantity');
+    table.dropColumn('product_quantity_unit');
+  });
 
-  await knex.schema.alterTable('soil_amendment_task', (table) => {
-    table.dropColumn('purpose');
-    table.renameColumn('temp_purpose', 'purpose');
+  // Alter cleaning task for volume and weight
+  await knex.schema.alterTable('pest_control_task', (table) => {
+    table.dropColumn('product_quantity');
+    table.dropColumn('product_quantity_unit');
+  });
+
+  // Alter product to add product state type id
+  // Removing npk to add new table instead no attempt to keep values as it is currently unused
+  await knex.schema.alterTable('product', (table) => {
+    table.dropChecks(['npk_unit_check', 'npk_percent_check']);
+    table.dropColumn('n');
+    table.dropColumn('p');
+    table.dropColumn('k');
+    table.dropColumn('npk_unit');
   });
 
-  await knex.schema.raw('DROP TYPE purpose_enum');
+  //Add permissions
+  // Use task or product permissions as needed
+};
+
+export const down = async function (knex) {
+  // No prod data unreleased feature - destroys data
+  await knex.schema.alterTable('product', (table) => {
+    table.decimal('n');
+    table.decimal('p');
+    table.decimal('k');
+    table.enu('npk_unit', ['ratio', 'percent']);
+    table.check(
+      '(COALESCE(n, p, k) IS NULL AND npk_unit IS NULL) OR (COALESCE(n, p, k) IS NOT NULL AND npk_unit IS NOT NULL)',
+      [],
+      'npk_unit_check',
+    );
+    table.check(
+      "npk_unit != 'percent' OR (npk_unit = 'percent' AND (n + p + k) <= 100)",
+      [],
+      'npk_percent_check',
+    );
+  });
 
   // Reverse product data migration
   await knex.schema.alterTable('soil_amendment_task', (table) => {
     table.decimal('product_quantity', 36, 12);
-    table.enu('product_quantity_unit', productQuantityUnits);
+    table.text('product_quantity_unit');
```

Comment:
Why text if it was an enum originally?

Reply:
I thought I checked that it is currently text so it should go back to that. Will double check!

Reply:
So I am seeing it as currently text on beta defaulting to 'kg' (which I think I forgot here):
![Screenshot 2024-05-29 at 9 18 33 AM](https://github.com/LiteFarmOrg/LiteFarm/assets/30075973/fd77cb13-735d-4e2c-aacf-3caace21f640)

Reply:
So you can see in the migration that created the table that they are both enumerations, but that's not a datatype -- you'll have to go to `constraints` > check (in the screenshot you have above) to see the list of allowable values.

I don't know how to create this view in pgAdmin4, but in Postico it's easiest to see in the DDL:

![Screenshot 2024-05-29 at 7 59 27 AM](https://github.com/LiteFarmOrg/LiteFarm/assets/82857964/a5013968-9c06-40d2-808d-89d4eb817505)

Reply:
Ah I def thought enum was a type, thanks! Will fix!